.teamcity/Gradle_Check/projects/FunctionalTestProject.kt

```diff
@@ -39,3 +31,15 @@ class FunctionalTestProject(model: CIBuildModel, testConfig: TestCoverage, stage
         }
     }
 }
+
+fun Project.configureBuildType(model: CIBuildModel, testConfig: TestCoverage, stage: Stage, subProject: GradleSubproject) {
+    val useDaemon = subProject.useDaemon && testConfig.testType != TestType.noDaemon
+
+    if (subProject.unitTests && testConfig.testType.unitTests) {
```

Comment:
I guess the whole if/else thing can be simplified by adding two methods to `GradleSubproject`:
```
fun hasTestsFor(type: TestType) = (unitTests && type.unitTests)
            || (functionalTests && type.functionalTests)
            || (crossVersionTests && type.crossVersionTests)

fun useDaemon(type: TestType) = useDaemon && (type != TestType.noDaemon)
```

Then the code simplifies to:
```
                        if (subProject.hasTestsOf(testConfig.testType)) {
                            buildType(FunctionalTest(model, testConfig, subProject.name, subProject.useDaemon(testConfig.testType), stage))
                        }

```

The we don't need this method any more.

Reply:
Yeah, brilliant idea.